limit.go

```diff
@@ -0,0 +1,145 @@
+package zoekt
+
+import "log"
+
+// SortAndTruncateFiles is a convenience around SortFiles and
+// DisplayTruncator. Given an aggregated files it will sort and then truncate
+// based on the search options.
+func SortAndTruncateFiles(files []FileMatch, opts *SearchOptions) []FileMatch {
+	SortFiles(files)
+	truncator, _ := NewDisplayTruncator(opts)
+	files, _ = truncator(files)
+	return files
+}
+
+// DisplayTruncator is a stateful function which enforces Document and Match
+// display limits by truncating and mutating before. hasMore is true until the
+// limits are exhausted. Once hasMore is false each subsequent call will
+// return an empty after and hasMore false.
+type DisplayTruncator func(before []FileMatch) (after []FileMatch, hasMore bool)
+
+// NewDisplayTruncator will return a DisplayTruncator which enforces the limits in
+// opts. If there are no limits to enforce, hasLimits is false and there is no
+// need to call DisplayTruncator.
+func NewDisplayTruncator(opts *SearchOptions) (_ DisplayTruncator, hasLimits bool) {
+	docLimit := opts.MaxDocDisplayCount
+	docLimited := docLimit > 0
+
+	matchLimit := opts.MaxMatchDisplayCount
+	matchLimited := matchLimit > 0
+
+	done := false
+
+	if !docLimited && !matchLimited {
+		return func(fm []FileMatch) ([]FileMatch, bool) {
+			return fm, true
+		}, false
+	}
+
+	return func(fm []FileMatch) ([]FileMatch, bool) {
+		if done {
+			return nil, false
+		}
+
+		if docLimited {
+			if len(fm) >= docLimit {
+				done = true
+				return fm[:docLimit], false
+			}
+			docLimit -= len(fm)
+		}
+
+		if matchLimited {
+			fm, matchLimit = limitMatches(fm, matchLimit, opts.ChunkMatches)
+			if matchLimit <= 0 {
+				done = true
+				return fm, false
+			}
+		}
```

Comment:
Yeah, so the logic here is wrong @keegancsmith.  When `docLimited && matchLimited`, we are only applying doc limits, even though match limits could take us to even fewer FileMatches.  Can't return until both have been applied.

Reply:
The test case, of course, only has matchLimits 🌞.

Reply:
#632